packages/mui-base/CONTRIBUTING.md

```diff
@@ -0,0 +1,180 @@
+# Contributing
+
+## Creating a new hook
+
+### File structure
+
+When creating a new hook, follow the file structure found in other hooks' directories:
+
+Taking the imaginary `useAwesomeControl` as an example:
+
+- 📂 `useAwesomeControl`
+  - 📄 `index.ts` - aggregates the public exports from all the other files in the directory
+  - 📄 `useAwesomeControl.test.tsx` - unit tests
+  - 📄 `useAwesomeControl.spec.tsx` - (optional) type tests
+  - 📄 `useAwesomeControl.ts` - the implementation of the hook
+  - 📄 `useAwesomeControl.types.ts` - type definitions
+
+### Implementation
+
+Most Base UI hooks have a similar structure:
+
+1. [Parameters destructuring](#parameters-destructuring)
+2. Hook-specific logic
+3. [Event handler factory functions](#event-handler-factory-functions)
+4. [Slot props resolvers](#slot-props-resolvers)
+
+#### Parameters destructuring
+
+The parameters type must be called `[HookName]Parameters`.
+There are docs generation scripts that require this pattern.
+
+#### Event handler factory functions
+
+We don't define event handlers directly as functions because they must be able to access and call other handlers provided by developers in slot props resolvers.
+
+In other words, instead of defining the `handleClick(event: React.MouseEvent) => void` function, we define the `createHandleClick(otherHandlers: EventHandlers) => (event: React.MouseEvent) => void`. The `otherHandlers` parameter contains external handlers provided by developers.
+
+By convention, we call them `createHandle[EventName]`.
+
+If we allow a developer to skip running our logic for a given handler, we check the `event.defaultMuiPrevented` field. It's an equivalent of the native `defaultPrevented` that works just for Base UI code:
+
+```tsx
+const createHandleKeyDown = (otherHandlers: EventHandlers) => (event: React.KeyboardEvent & MuiCancellableEvent) => {
+  // Run the external handler first.
+  // It possibly can set the defaultMuiPrevented field.
+  otherHandlers.onKeyDown?.(event);
+
+  // If the field is set, do not execute the usual handler logic.
+  if (event.defaultMuiPrevented) {
+    return;
+  }
+
+  // handler-specific logic...
+```
+
+#### Slot props resolvers
+
+These are functions called `get[SlotName]Props` that accept additional (optional) event handlers and return props to be placed on slots of a component.
+Many components have just one slot (that we call "root"), but more complex components can have more.
+
+The order of merging props for the resulting object is important so that users can override the build-in props when necessary:
+
+1. built-in props
+2. external props
+3. ref
+4. event handlers
+
+Refs and event handlers can be placed in any order.
+They just have to be after external props.
+
+Example:
+
+```tsx
+const getRootProps = <OtherProps extends EventHandlers>(
```

Comment:
Why is the parameter named `otherProps` here and `otherHandlers` on the next example?

Reply:
Ah, sorry, my bad. I should have made it consistent.
Although it actually reflects the reality - we used to accept only event handlers in this parameter, but we're changing this to accept arbitrary props (#38186). So ultimately, it will be `otherProps`.

Reply:
OK
This is clearly a small issue :laughing: